### ISSUE TITLE
Feature Request

### ISSUE THREAD

**Issue by gaithoben** [arangodb/arangodb#14387]
## My Environment

* __ArangoDB Version__:        
* __Storage Engine__:             
* __Deployment Mode__:       
* __Deployment Strategy__:   
* __Configuration__:               
* __Infrastructure__:               
* __Operating System__:        
* __Total RAM in your machine__:        
* __Disks in use__:        
* __Used Package__:              

## Component, Query & Data

This is a feature request, maybe it's in the roadmap, please correct me. Is it possible to have a [ref datatype as it is in faundadb? ](https://docs.fauna.com/fauna/current/api/fql/functions/ref?lang=javascript). In short, a pointer to a document in another collection or the same collection.

Thank you for the wonderful job.

**Comment by dothebart** [comment]
this can basically be achieved by putting  `collection/document_key` into a datafield (similar to the `_from` and `_to` fields in edge collections), and maybe adding it to the index used by the queries you invoke, so you can use projections to save you from reading the document itself (if you don't need its content).

You can then feed this value into the AQL [`DOCUMENT()` function](https://www.arangodb.com/docs/stable/aql/functions-miscellaneous.html#document) to fetch the actual document. 

You could also use [joins](https://www.arangodb.com/docs/stable/aql/examples-join.html) filtering for `_key`.

**Comment by gaithoben** [comment]
Thank you so much @dothebart.

Is it possible to enforce referential integrity such that a referenced document must always exist?

**Comment by dothebart** [comment]
hm, you'd probably either write a foxx service that ensures this during inserting, or if you are inserting the documents at once use a transaction for that.

Graph integrity is only waranted at a similar level - if you use document operations instead of the graph API to insert / delete edges / vertices, this also isn't waranted.

**Comment by gaithoben** [comment]
Thank you very much!

I hope in the future we can get a referential integrity warranty.

**Comment by coronabytes** [comment]
Hi,

It would be really cool if DOCUMENT() would no longer be required.
Because the database "knows" it's a referenced document field, data access becomes
```
FOR x in projects
RETURN x.client.name
```
instead of
```
FOR x in projects
RETURN DOCUMENT(clients, x.clientKey).name
```

and then make the indices smart enought to be able to FILTER and SORT on a referenced field
WITHOUT performance penalty (now denormalization is the only way if one likes fast queries)
Basically allow indices into reference fields (client.name on projects collection)

from what I heard OrientDB can do something like that

**Comment by gaithoben** [comment]
I think what I am asking for is, If we have an edge connecting two documents A & B, Neither A nor B can be deleted without having to delete the edge first. Because the edge is the "RELATIONSHIP".

**Comment by coronabytes** [comment]
Yup, you only want references with "Foreign Key Constraints" and as the go hand in hand with indices in SQL I'd propose some performance optimizations based on them.
